text/0000-promise.md

```diff
@@ -0,0 +1,460 @@
+    }
+}
+
+trait Context<'a> {
+    fn task<'cx, O, E>(&'cx mut self, execute: E) -> TaskBuilder<Self, E>
+        where
+            'a: 'cx,
+            O: Send + 'static,
+            E: FnOnce() -> O + Send + 'static,
+    {
+        TaskBuilder::new(self, execute)
+    }
+}
+```
+
+Internally, `TaskBuilder` uses [`napi_async_work`](https://nodejs.org/api/n-api.html#n_api_simple_asynchronous_operations) for creating and scheduling tasks.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The most significant drawback is that the user is presented with many more options for asynchronous programming without clear direction on which to use. Neon will need to lean on documentation to guide users to the correct APIs for various situations. 
+
+## Rationale and alternatives
+[alternatives]: #alternatives
+
+### Common `Deferred` pattern
+
+Most promise libraries that provide a `Deferred` object provide the `promise` as part of that object. Neon might have a similar approach:
+
+```rust
+struct Deferred<'a> {
+    handle: *mut c_void,
+    promise: Handle<'a, JsPromise>,
+}
+```
+
+The issue with this approach is that `Deferred` is `Send` and `JsPromise` is *not*. We would also need to provide a getter for the resolve/reject handle. That getter would need to _consume_ self because it cannot be used multiple times. This would result in worse ergonomics.
+
+Instead a tuple is returned, similar to `(sender, receiver)` returned from `std::sync::mpsc::channel`.
+
+### Low level async work
+
+Neon could provide a lower level `AsyncWork` or `trait` that only accepts static function pointers and `&mut data` instead of consuming data. However, this would push complexity down for threading consumable state, try/catch and several other subtle issues. It is also boilerplate heavy.
+
+### Catching panics
+
+A `panic` in a `TaskBuilder::promise` will drop the promise without resolution. We could catch the unwind and convert to a rejection; however, catching panics is not considered best practice. Since we do not require catching unwind to prevent unsound unwinding across FFI, it's better to have a course grained control where the `Deferred` rejects when it drops without being settled.
+
+## Future Expansion
+
+Node-API allows async work to be [cancelled](https://nodejs.org/api/n-api.html#n_api_napi_cancel_async_work) after it has been scheduled, but before it has executed. This is left as future improvement since we do not have a clear use case in mind and it can be added without breaking changes.
+
+Two patterns for adding it without breaking change:
+* Add `fn handle(&self) -> TaskHandle` to `TaskBuilder` to allow creating cancellation handles before scheduling
+* Add methods that return a cancellation handle in addition to the thunk (e.g., `fn cancellable_promise(self, f: F) -> (TaskHandle, Handle<JsPromise>)`
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Are the thunk names `TaskBuilder::promise` and `TaskBuilder::complete` clear?
```

Comment:
IMO `TaskBuilder::promise` is very clear.

Re: `complete`, a few possible alternatives:
- `on_complete`
- `and_then` (by analogy to various Rust std APIs)

Not sure what I think, but `complete` doesn't seem bad.

Reply:
I like `and_then` since it matches common naming in Rust, for example, in the [`futures` crate](https://docs.rs/futures/0.3.17/futures/future/trait.TryFutureExt.html#method.and_then).